akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -394,6 +396,62 @@ final case class Fold[In, Out](zero: Out, f: (Out, In) ⇒ Out) extends GraphSta
 /**
  * INTERNAL API
  */
+final class FoldAsync[In, Out](zero: ⇒ Out, f: (Out, In) ⇒ Future[Out]) extends GraphStage[FlowShape[In, Out]] {
+  import akka.dispatch.ExecutionContexts
+
+  val in = Inlet[In]("FoldAsync.in")
+  val out = Outlet[Out]("FoldAsync.out")
+
+  val shape = FlowShape.of(in, out)
+
+  override def toString: String = "FoldAsync"
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+      private var aggregated: Out = zero
+      private var aggregating: Future[Out] = Future.successful(aggregated)
+
+      private def handle(output: Try[Out]): Unit = output match {
+        case Failure(reason) ⇒ fail(out, reason)
+        case Success(o) ⇒
+          aggregated = o
+          if (!isClosed(in)) pull(in)
+      }
+
+      private def ec = ExecutionContexts.sameThreadExecutionContext
+
+      setHandler(in, new InHandler {
+        val asyncCallback = getAsyncCallback[Try[Out]](handle).invoke _
+
+        def onPush(): Unit = {
+          aggregating = f(aggregated, grab(in))
+          aggregating.onComplete(asyncCallback)(ec)
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          val onTermination: Try[Out] ⇒ Unit = {
+            case Failure(reason) ⇒ failStage(reason)
+            case Success(folded) ⇒
+              emit(out, folded)
+
+              completeStage()
+          }
+          val watchTermination =
+            getAsyncCallback[Try[Out]](onTermination).invoke _
+
+          aggregating.onComplete(watchTermination)(ec)
```

Comment:
Take a look at how `MapAsyncUnordered` contains an optimization for the future result to avoid scheduling the callback on the dispatcher if the value is already here.